Differentiable parameter-shift gradient transform

josh146 opened the pull request:
**Context:** As part of the roadmap for supporting differentiable batch execution, gradient logic will be moved out of the subclasses and into a module of pure functions. This is the second PR after #1476; here, we move the finite-difference logic out of `QubitParamShiftTape` and into a new gradients package.

**Description of the Change:**

- Adds a function `expval_param_shift` for computing the parameter-shift gradient of a tape terminating in expectation values. Directly equivalent to `QubitParamShift.parameter_shift`.

- Adds a function `var_param_shift` for computing the parameter-shift gradient of a tape terminating in one or more variances. Directly equivalent to `QubitParamShift.parameter_shift_var`.

- Adds a wrapper function `param_shift`, which does the following:

  - Basic input validation
  - Performs static analysis of the tape parameters, to determine which support parameter-shift rules.
  - A fallback mode; `finite_diff` is called for any unsupported parameters
  - Dispatches to one of `expval_param_shift` or `var_param_shift` depending on the structure of the tape
  - Provides a processing function to combine the gradients computed via fallback and the parameter-shift methods.

**Benefits:**

- The parameter-shift logic is now much more user and dev accessible
- The parameter-shift logic is now differentiable, allowing higher-order derivatives to be accessed no matter the gradient recipe.
- Redundant and zero terms are automatically removed from gradient recipes
- If the output value of the unshifted, input tape is known, and the gradient recipe contains an unshifted component, this can be provided to the parameter-shift rule to reduce the number of evaluations required.

**Possible Drawbacks:**

- See drawbacks in #1476 
- For variance support, the only non-involutory observable supported is `qml.Hermitian`. We should consider extending support to `qml.Hamiltonian`, now that `expval(H)` is supported.
- We require `JacobianTape` as input, since we continue to rely on that subclasses static gradient analysis methods
- To cache the static gradient analysis, it is appended to the input tape - a better approach should be considered.

**Related GitHub Issues:**

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+        gradient_tapes.extend(pdA2_tapes)
+
+    def processing_fn(results):
+        mask = qml.math.convert_like(qml.math.reshape(var_mask, [-1, 1]), results[0])
```
Not sure I'm grasping this line :thinking:

josh146 replied:
Ah, two things are happening here:

1. We need to expand the dimensions of the variance mask
2. Next, we need to convert it to be the same type as the results. E.g., if the results is a Torch tensor, `var_mask` must also be a torch tensor.

antalszava replied:
Thank you! It can be worth noting this as a comment or adjusting names to reflect these steps.

josh146 replied:
Have added a comment to the file!

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    # evaluate the analytic derivative of <A>
+    pdA_tapes, pdA_fn = expval_param_shift(expval_tape, argnum, shift, gradient_recipes, f0)
+    gradient_tapes.extend(pdA_tapes)
+    tape_boundary = len(pdA_tapes) + 1
```
Wasn't sure what `tape_boundary` was, when it was used, how about its assignment down a bit?

josh146 replied:
Let me add a comment here.

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+                m = qml.math.convert_like(m, pdA2)
+                pdA2 = qml.math.where(qml.math.reshape(m, [-1, 1]), 0, pdA2)
+
+        # return d(var(A))/dp = d<A^2>/dp -2 * <A> * d<A>/dp for the variances,
+        # d<A>/dp for plain expectations
+        return qml.math.where(mask, pdA2 - 2 * f0 * pdA, pdA)
```
When would we return plain expectations at this point?

josh146 replied:
`mask` is `True` for `var()`, and `False` for `expval()`. I'll update the comment to note this.

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+            pdA2 = pdA2_fn(results[tape_boundary:])
+
+            if involutory:
+                # if involutory observables are present, ensure they have zero gradient
+                m = [tape.observables[i].name != "Hermitian" for i in var_idx]
+                m = qml.math.convert_like(m, pdA2)
+                pdA2 = qml.math.where(qml.math.reshape(m, [-1, 1]), 0, pdA2)
```
Before this logic, what does `pdA2` contain for involutory observables?

josh146 replied:
For the `pdA2_tapes`, we have replaced non-involutory observables with their square (`A` -> `A^2`). However, involutory observables have been left as-is; as a result, they will give the incorrect result! Hence why we need to replace the value with 0 (the known, correct result for involutory variables).

josh146 replied:
Have added a bigger comment here :)

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+            trainable indices are returned.
+        shift (float): The shift value to use for the two-term parameter-shift formula.
+            Only valid if the operation in question supports the two-term parameter-shift
+            rule (that is, it has two distinct eigenvalues) and ``gradient_recipe``
```
```suggestion
            rule (that is, it has two distinct eigenvalues) and ``gradient_recipes``
```

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    return gradient_tapes, processing_fn
+
+
+def param_shift(
```
This function doesn't seem to appear in the docs, would that come up as a later addition?

josh146 replied:
Yep, I am cogent over how big this PR is, and thought it might make sense to have a final PR at the end of this series just for the docs :)

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    ...     qml.expval(qml.PauliZ(0))
+    ...     qml.var(qml.PauliZ(0))
+    >>> tape.trainable_params = {0, 1, 2}
+    >>> gradient_tapes, fn = qml.gradients.param_shift(tape)
+    >>> res = dev.batch_execute(gradient_tapes)
+    >>> fn(res)
+    [[-0.38751721 -0.18884787 -0.38355704]
+     [ 0.69916862  0.34072424  0.69202359]]
```
:heart_eyes:

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+
+    # perform gradient method validation
+    if any(m.return_type is qml.operation.State for m in tape.measurements):
+        raise ValueError("Does not support circuits that return the state")
```
```suggestion
        raise ValueError("Computing the gradient of circuits that return the state is not supported.")
```

antalszava replied:
How about `qml.sample`? Should we raise error for both (and more if applicable) at the same part of the code base?

josh146 replied:
@antalszava I might leave this one as-is, since it matches the existing error message?

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    _gradient_analysis(tape)
+    gradient_tapes = []
+
+    # TODO: replace the JacobianTape._grad_method_validation
```
What needs to be replaced?

josh146 replied:
The entire method, as we plan to deprecate `JacobianTape` :)

antalszava replied:
Hmm, I think I still don't fully grasp the comment. Would we not deprecate `JacobianTape` as is and `JacobianTape._grad_method_validation` with it? How come we'd need to revisit this particular spot?

josh146 replied:
On line 468, the `JacobianTape._grad_method_validation` method is being called :) Just leaving a todo here so that we know it needs to be replaced once `JacobianTape` is deprecated

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    # functionality before deprecation.
+    method = "analytic" if fallback_fn is None else "best"
+    diff_methods = tape._grad_method_validation(method)
+
+    if not tape.trainable_params or all(g == "0" for g in diff_methods):
+        # Either all parameters have grad method 0, or there are no trainable
+        # parameters.
```
Maybe we could save a comment here?
```suggestion
    all_params_grad_method_zero = all(g == "0" for g in diff_methods)
    if not tape.trainable_params or all_params_grad_method_zero:
```

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    if not tape.trainable_params or all(g == "0" for g in diff_methods):
+        # Either all parameters have grad method 0, or there are no trainable
+        # parameters.
+        return gradient_tapes, lambda x: np.zeros([tape.output_dim, len(tape.trainable_params)])
```
```suggestion
        return gradient_tapes, lambda _: np.zeros([tape.output_dim, len(tape.trainable_params)])
```

josh146 replied:
nice, I didn't know you could do this!

antalszava commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,848 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
```
```suggestion
# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
```

antalszava commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,848 @@
+        assert tape._par_info[0]["grad_method"] == "A"
+        assert tape._par_info[1]["grad_method"] == "0"
+
+        # in non-graph mode, it is impossible to determine
+        # if a parameter is independent or not
+        tape._gradient_fn = None
+        tape._graph = None
+        _gradient_analysis(tape, use_graph=False)
+
+        assert tape._par_info[1]["grad_method"] == "A"
```
This seems to be a separate mini test case.

josh146 replied:
I'll separate it out into a new test!

antalszava commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,848 @@
+        assert tape._par_info[2]["grad_method"] == "A"
+
+
+def grad_fn(tape, dev, fn=qml.gradients.param_shift, **kwargs):
```
Worth adding a short docstring.

josh146 replied:
```suggestion
def grad_fn(tape, dev, fn=qml.gradients.param_shift, **kwargs):
    """Utility function to automate execution and processing of gradient tapes"""
```

antalszava commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,848 @@
+        assert res.shape == (1, 2)
+
+        # only called for parameter 0
+        assert spy.call_args[0][0:2] == (tape, [0])
```
Should there be an `assert` here to further check that `spy.call_args` doesn't contain other elements?

josh146 replied:
Don't think so; here, I am only checking the first two elements :) I don't worry about what the others are (there are definitely more than one!)

antalszava replied:
Oh right! So the first two elements are checked for `spy.call_args[0]`, though would we want to make sure that there's no `spy.call_args[1]` here? Or that the function was called only once as per `# only called for parameter 0`.

josh146 replied:
> though would we want to make sure that there's no spy.call_args[1] here?

Nope, we are only interested in checking the positional arguments, not the keyword arguments :)

> Or that the function was called only once as per # only called for parameter 0.

This is checked via the RHS; we are ensuring that the function is being called with `tape=tape, argnum=[0]`.

antalszava commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,848 @@
+
+        fn(dev.batch_execute(tapes))
+
+    def test_independent_parameters(self):
```
Had a `test_independent_parameter` test case before, worth somehow describing how this case differs.
```suggestion
    def test_independent_parameters_analytic(self):
```

josh146 replied:
Oops! good catch!

antalszava commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,848 @@
+    """Test that the transform is differentiable"""
+
+    def test_autograd(self, tol):
+        """Tests that the output of the finite-difference transform
```
Repeats for other test cases too.
```suggestion
        """Tests that the output of the parameter-shift transform
```

antalszava commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,848 @@
+        config.update("jax_enable_x64", True)
+
+        dev = qml.device("default.qubit", wires=2)
+        params = jnp.array([0.543, -0.654])
```
Are these parameters marked as differentiable for jax this way? :thinking:

josh146 replied:
Yep! `jnp` arrays are always differentiable.

antalszava reviewed (commented):
Hi @josh146, having the functional approach will be awesome! :heart_eyes:  :+1: Some names may be nice to change to make some parts more readable and curious about `qml.Projector` as another non-involutory observable. The remaining comments were minor/questions.

josh146 commented on pennylane/gradients/__init__.py:
```diff
@@ -15,5 +15,7 @@
+from . import parameter_shift
 
 from .finite_difference import finite_diff, finite_diff_stencil, generate_shifted_tapes
+from .parameter_shift import param_shift
```
Ah! This is so that it can be used like
```python
>>> qml.gradients.param_shift(tape)
```

:) 

Which I think is quite clear

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+
+
+def _get_operation_recipe(tape, t_idx, shift=np.pi / 2):
+    """Utility function to return the parameter-shift rule
```
Yep! This is deliberate, as per PEP8:

> Write docstrings for all public modules, functions, classes, and methods. Docstrings are not necessary for non-public methods, but you should have a comment that describes what the method does. This comment should appear after the "def" line.

It is basically up to the code author and reviewers how detailed the docstring of a private method should be. I typically like a middling approach - a docstring that describes what the private method/function does, but isn't nitpicky about following our standard docstring conventions.

antalszava replied:
Oh thanks, that's nice to know :)

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    If the corresponding operation has grad_recipe=None, then
+    the default two-term parameter-shift rule is assumed.
+    """
+    p_idx = list(tape.trainable_params)[t_idx]
```
done! Thanks Antal. Have also added comments here.

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    """
+    p_idx = list(tape.trainable_params)[t_idx]
+    op = tape._par_info[p_idx]["op"]
+    op_p_idx = tape._par_info[p_idx]["p_idx"]
```
Yep! have made this more clear.

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+
+def _process_gradient_recipe(gr, tol=1e-10):
+    """Utility function to process gradient recipes."""
+    gr = np.array(gr).T
```
done!

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    return gr[:, np.argsort(np.abs(gr)[-1])]
+
+
+def _gradient_analysis(tape, use_graph=True):
```
Potentially, but `use_graph` is by default True in `QubitParamShiftTape`, and I don't want to change user-facing behaviour

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+            rule (that is, it has two distinct eigenvalues) and ``gradient_recipe``
+            is ``None``.
+        gradient_recipes (tuple(list[list[float]] or None)): List of gradient recipes
+            for the parameter-shift method. One gradient recipe must be provided
```
It is not being checked, _but_ the `for` loop will fail if this is not the case.

Due to performance, I try to avoid validating everything - I hope that the docstrings will suffice is most cases! Let me know what you think

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+        gradient_recipes (tuple(list[list[float]] or None)): List of gradient recipes
+            for the parameter-shift method. One gradient recipe must be provided
+            per trainable parameter.
+        f0 (tensor_like[float] or None): Output of the evaluated input tape. If provided,
```
This one I might argue to keep as is, since it corresponds to the mathematical notation we commonly see ($f_0$)

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    gradient_tapes = []
+    gradient_coeffs = []
+    shapes = []
+    c0 = []
```
done!

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+            continue
+
+        # get the gradient recipe for the trainable parameter
+        gr = gradient_recipes[argnum.index(idx)]
```
done!

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+        start = 1 if c0 and f0 is None else 0
+        r0 = f0 or results[0]
+
+        for i, s in enumerate(shapes):
```
I feel like this one is pretty clear in context though?

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+        # The following is for backwards compatibility; currently,
+        # the device stacks multiple measurement arrays, even if not the same
+        # size, resulting in a ragged array.
+        # In the future, we might want to change this so that only tuples
+        # of arrays are returned.
```
The reason I avoid making this a TODO, is it is a breaking change that requires discussion first :)

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,450 @@
+    tape_boundary = len(pdA_tapes) + 1
+
+    # For involutory observables (A^2 = I) we have d<A^2>/dp = 0.
+    # Currently, the only observable we have in PL that may be non-involutory is qml.Hermitian
```
`qml.Projector` should be involutory as well (correct me if wrong!)

glassnotes replied:
Projectors are non-involutory, they square to themselves `(|A><A| |A><A| = |A><A|)`.

josh146 replied:
🤦 This is why I shouldn't try and matrix multiply in my head past 9pm. Thanks @glassnotes

josh146 replied:
I'll try and sneak in projector support.

glassnotes commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,471 @@
+
+import pennylane as qml
+
+from .finite_difference import finite_diff, generate_shifted_tapes
```
Depending on how many more such transforms there will be, would it be worth moving something like `generate_shifted_tapes` into a `utils` file?

josh146 replied:
Yes, I was thinking the same! I think so.

glassnotes commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,471 @@
+    return op.get_parameter_shift(op_p_idx, shift=shift)
+
+
+def _process_gradient_recipe(gradient_recipe, tol=1e-10):
```
It looks like similar logic is used in the `finite_diff_coeffs` so might be something else for a utils module? (but only if it makes sense, and reuse would be possible)

josh146 replied:
Yes, the only difference is that parameter-shift rules can _scale_ the parameter, which finite-differences doesn't.

glassnotes commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,471 @@
+
+def var_param_shift(tape, argnum, shift=np.pi / 2, gradient_recipes=None, f0=None):
+    r"""Generate the parameter-shift tapes and postprocessing methods required
+    to compute the gradient of an gate parameter with respect to a
```
```suggestion
    to compute the gradient of a gate parameter with respect to a

```

glassnotes reviewed (commented):
Have looked at everything but the tests, but not in extensive mathematical detail. Will continue later this morning!

josh146 commented:
@antalszava @glassnotes: I fixed the variance rule to take into account Projectors, and also added a test, in commits https://github.com/PennyLaneAI/pennylane/pull/1479/files/be46c0ab7019b80cfae4f932c41a2c8028391191..0c5666c7bdb3bb390ca967660d272953973eed1b

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,514 @@
+    allowing us to compute the gradient using :math:`2N + 1` evaluations.
+
+    **Example**
+
```
Oh one note here: could be good to have some `params` to have the example be executable easily after copying.

:thinking: 

```python
import pennylane as qml
from pennylane import numpy as np

params = np.array([0.1, 0.2, 0.3], requires_grad=True)

with qml.tape.QuantumTape() as tape:
    qml.RX(params[0], wires=0)
    qml.RY(params[1], wires=0)
    qml.RX(params[2], wires=0)
    qml.expval(qml.PauliZ(0))
    qml.var(qml.PauliZ(0))
    
tape.trainable_params = {0, 1, 2}
gradient_tapes, fn = qml.gradients.param_shift(tape)
res = dev.batch_execute(gradient_tapes)
fn(res)
```
```
~/pennylane/pennylane/gradients/parameter_shift.py in _gradient_analysis(tape, use_graph)
    112                 info["grad_method"] = "F"
    113             else:
--> 114                 info["grad_method"] = tape._grad_method(
    115                     idx, use_graph=use_graph, default_method="A"
    116                 )

AttributeError: 'QuantumTape' object has no attribute '_grad_method'
```

josh146 replied:
Good catch!

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -23,6 +23,40 @@
+                components_squared.append(NONINVOLUTORY_OBS[comp.name](comp))
+            except KeyError:
+                # component is involutory
+                pass
```
Looks like this line is not being covered.

josh146 replied:
Added a test for this!

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -23,6 +23,40 @@
 from .finite_difference import finite_diff, generate_shifted_tapes
 
 
+NONINVOLUTORY_OBS = {
+    "Hermitian": lambda obs: obs.__class__(obs.matrix @ obs.matrix, wires=obs.wires),
+    "SparseHamiltonian": lambda obs: obs.__class__(obs.matrix @ obs.matrix, wires=obs.wires),
+    "Projector": lambda obs: obs,
+}
```
:100: It's great to have this place in case the list will further grow.

josh146 replied:
Ideally, we won't have this in the future! Once the operator refactor is done, hopefully you should just be able to do `obs = obs @ obs` :)

antalszava commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -719,6 +723,42 @@ def test_expval_and_variance(self, tol):
+        expected = 0.25 * np.sin(x / 2) ** 2 * (3 + np.cos(2 * y) + 2 * np.cos(x) * np.sin(y) ** 2)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # # circuit jacobians
```
How about making this its own test case? It would make sure that no state is being left when calling `tape.execute` and would separate the cases better. Costs additional lines though.

josh146 replied:
I would lean to keeping it as is, partly because `dev.execute` should automatically reset between calls, and partly because it matches the structure of the other tests.

antalszava commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,514 @@
+
+NONINVOLUTORY_OBS = {
+    "Hermitian": lambda obs: obs.__class__(obs.matrix @ obs.matrix, wires=obs.wires),
+    "SparseHamiltonian": lambda obs: obs.__class__(obs.matrix @ obs.matrix, wires=obs.wires),
```
Oh, just realized it seems like variance does not apply for [SparseHamiltonian at the moment](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/ops/qubit.py#L3255):
```
    .. warning::
        ``SparseHamiltonian`` observables can only be used to return expectation values.
        Variances and samples are not supported.
```

Maybe worth still leaving this as is once the support is in? :thinking:

josh146 replied:
Yep, I left it here just for the future :)

antalszava reviewed (commented):
Just have some questions/suggestions, otherwise it's looking awesome! :tada: 

After our chat yesterday, curious: how would we handle template/tape expansions?

As you've suggested, the following now works like a charm:
```python
n_layers = 4
n_wires = 4

dev = qml.device('default.qubit', wires=n_wires)

rng = np.random.default_rng(seed=42)
params = np.array(rng.standard_normal((n_layers, n_wires)))

with qml.tape.JacobianTape() as tape:
    qml.templates.BasicEntanglerLayers(params, wires=range(n_wires)).expand()
    qml.expval(qml.PauliZ(0))
    
tape = tape.expand(stop_at=lambda obj: not isinstance(obj, qml.tape.QuantumTape) and dev.supports_operation(obj.name))

tape.trainable_params = set(range(n_layers * n_wires))
```
Not sure, however, if it will be intuitive to users to call `expand` on the template and then call `tape.expand` too. Maybe it's for another PR?

josh146 commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,514 @@
+    allowing us to compute the gradient using :math:`2N + 1` evaluations.
+
+    **Example**
+
+    >>> with qml.tape.QuantumTape() as tape:
```
```suggestion

    >>> params = np.array([0.1, 0.2, 0.3])
    >>> with qml.tape.QuantumTape() as tape:
```

glassnotes commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,514 @@
+    Returns:
+        tuple[list[QuantumTape], function]: A tuple containing a
+        list of generated tapes, in addition to a post-processing
+        function to be applied to the evaluated tapes.
```
```suggestion
        function to be applied to the results of the evaluated tapes.
```

josh146 commented:
> Not sure, however, if it will be intuitive to users to call expand on the template and then call tape.expand too. Maybe it's for another PR?

@antalszava yes exactly 🙂  This PR simply adds the low-level tape transform. A future PR will add support via QNodes, which is always the user-facing level in PL.

glassnotes commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,514 @@
+    tape, argnum=None, shift=np.pi / 2, gradient_recipes=None, fallback_fn=finite_diff, f0=None
+):
+    r"""Generate the parameter-shift tapes and postprocessing methods required
+    to compute the gradient of an gate parameter with respect to an
```
```suggestion
    to compute the gradient of a gate parameter with respect to an
```

glassnotes commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,514 @@
+    Returns:
+        tuple[list[QuantumTape], function]: A tuple containing a
+        list of generated tapes, in addition to a post-processing
+        function to be applied to the evaluated tapes.
```
```suggestion
        function to be applied to the results of the evaluated tapes.
```

glassnotes commented on pennylane/gradients/parameter_shift.py:
```diff
@@ -0,0 +1,514 @@
+            If ``None``, the default gradient recipe containing the two terms
+            :math:`[c_0, a_0, s_0]=[1/2, 1, \pi/2]` and :math:`[c_1, a_1,
+            s_1]=[-1/2, 1, -\pi/2]` is assumed for every parameter.
+        fallback_fn (None or Callable): a fallback grdient function to use for
```
```suggestion
        fallback_fn (None or Callable): a fallback gradient function to use for
```

glassnotes commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,889 @@
+        assert tape._par_info[1]["grad_method"] == "A"
+        assert tape._par_info[2]["grad_method"] == "A"
+
+        _gradient_analysis(tape)
```
Why is this being run a second time?

josh146 replied:
Ah... because I am being sneaky and trying to sneak in extra lines for coverage 🙈 

The gradient analysis is cached, so this line simply catches the 'return cached results' section of the code. I'll write a proper test for this.

glassnotes commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,889 @@
+            qml.RY(-0.654, wires=[0])
+            qml.expval(qml.PauliZ(0))
+
+        gradient_recipes = [[[-1e7, 1, 0], [1e7, 1, 1e7]], [[-1e7, 1, 0], [1e7, 1, 1e7]]]
```
Why are these numbers so large? :fearful:

josh146 replied:
oh this is forward finite differences haha. So the coefficients are `±1/h=±1/1e-7 =1e7` 😆

glassnotes commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,889 @@
+
+    def test_independent_parameters_analytic(self):
+        """Test the case where expectation values are independent of some parameters. For those
+        parameters, the gradient should be evaluated to zero without executing the device."""
```
This is actually such an awesome feature :grin:

josh146 replied:
It's a contentious feature! Because it uses networkx which can be slow 😬 

It depends on your perspective; do you want to save quantum compute at the expense of classical compute?

glassnotes commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,889 @@
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_involutory_variance(self, tol):
+        """Tests qubit observable that are involutory"""
```
```suggestion
        """Tests qubit observables that are involutory"""
```

glassnotes commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,889 @@
+
+    def test_involutory_and_noninvolutory_variance(self, tol):
+        """Tests a qubit Hermitian observable that is not involutory alongside
+        a involutory observable."""
```
```suggestion
        an involutory observable."""

```

glassnotes commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -0,0 +1,889 @@
+        assert len(tapes) == 4
+
+        res = fn(dev.batch_execute(tapes))
+        assert res.shape == (5, 2)
```
I don't quite follow how this output shape is obtained, is it because there is 1 expectation value and 4 output probabilities, with a gradient for each of the two parameters?

josh146 replied:
Yep! The Jacobian is simply reshaped to be a 2D array, so the 1 expval and the 4 probs = 5 outputs. Coupled with the 2 parameters, the output Jacobian has shape `(5, 2)`.

glassnotes reviewed (commented):
Have made it through all the test cases, mostly just caught typos this time around.

One question I have is what happens to the gates that use the four-term shift rule, like the controlled rotations. They're included in some test cases, but the docs for `param_shift` are pretty clear about the shift value only being valid for the two-term rule, but the controlled rotation tests don't pass any custom parameters or gradient recipes.

glassnotes commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -57,12 +78,18 @@ def test_independent(self):
-        tape._graph = None
+    def test_independent_no_graph_mode(self):
+        """In non-graph mode, it is impossible to determine
+        # if a parameter is independent or not"""
```
```suggestion
        if a parameter is independent or not"""
```

glassnotes reviewed (approved):
Thanks @josh146 , looks like all the changes are in, happy to approve!

josh146 commented:
[ch7869]

antalszava commented on tests/gradients/test_parameter_shift.py:
```diff
@@ -413,13 +442,15 @@ def test_variance_gradients_agree_finite_differences(self, tol):
             qml.RX(params[2], wires=[0])
             qml.CNOT(wires=[0, 1])
-            qml.expval(qml.PauliZ(0)), qml.var(qml.PauliX(1))
+            qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(0) @ qml.PauliX(1))
```
How come this was changed?

josh146 replied:
I don't recall 🤔

antalszava reviewed (approved):
Looks good to me! :muscle: :heart_eyes: Excited for this!
